### PR TITLE
fix: multiple patches during long running tests for LMQ over RocksDB

### PR DESCRIPTION
**Pull request by lizhanhui**
### Which Issue(s) This PR Fixes

To #8829

### CODE REVIEW

File: store/src/main/java/org/apache/rocketmq/store/rocksdb/RocksDBOptionsFactory.java [hunk 1]
```diff
@@ -144,10 +146,8 @@ public static DBOptions createDBOptions() {
                 setInfoLogLevel(InfoLogLevel.INFO_LEVEL).
                 setWalRecoveryMode(WALRecoveryMode.PointInTimeRecovery).
                 setManualWalFlush(true).
-                setMaxTotalWalSize(0).
```
Comment (tianliuliu) [inline store/src/main/java/org/apache/rocketmq/store/rocksdb/RocksDBOptionsFactory.java:147]: 这里是多个cf使用这个db, 可以设置一个MaxTotalWalSize 来实现wal switch

File: store/src/main/java/org/apache/rocketmq/store/queue/RocksGroupCommitService.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+            if (null == dispatchRequest || requests.size() >= PREFERRED_DISPATCH_REQUEST_COUNT) {
+                while (!store.isStopped()) {
+                    try {
+                        // putMessagePosition will clear requests after consume queue building completion
+                        store.putMessagePosition(requests);
+                        break;
+                    } catch (RocksDBException e) {
+                        log.error("Failed to build consume queue in RocksDB", e);
+                    }
```
Comment (RongtongJin) [inline store/src/main/java/org/apache/rocketmq/store/queue/RocksGroupCommitService.java:91]: If throwing a RocksDBException leads to an endless loop without any pause,  is this expected behavior?

File: store/src/main/java/org/apache/rocketmq/store/queue/RocksDBConsumeQueueStore.java [hunk 3]
```diff
@@ -500,6 +529,16 @@ public ConsumeQueueInterface findOrCreateConsumeQueue(String topic, int queueId)
...
 
+    @Override
+    public ConcurrentMap<Integer, ConsumeQueueInterface> findConsumeQueueMap(String topic) {
+        if (MixAll.isLmq(topic)) {
+            ConcurrentMap<Integer, ConsumeQueueInterface> result = new ConcurrentHashMap<>(1);
+            result.put(MixAll.LMQ_QUEUE_ID, findOrCreateConsumeQueue(topic, MixAll.LMQ_QUEUE_ID));
+            return result;
+        }
+        return super.findConsumeQueueMap(topic);
+    }
```
Comment (RongtongJin) [inline store/src/main/java/org/apache/rocketmq/store/queue/RocksDBConsumeQueueStore.java:540]: I don't quite understand the necessity of overriding this method. Could you please clarify?

File: store/src/main/java/org/apache/rocketmq/store/rocksdb/RocksDBOptionsFactory.java (hunk 1 above)
Comment (lizhanhui) [reply store/src/main/java/org/apache/rocketmq/store/rocksdb/RocksDBOptionsFactory.java:147]: ConsumeQueue uses AtomicFlush and WAL is disabled, this option does not make sense here.

File: store/src/main/java/org/apache/rocketmq/store/queue/RocksGroupCommitService.java (hunk 2 above)
Comment (lizhanhui) [reply store/src/main/java/org/apache/rocketmq/store/queue/RocksGroupCommitService.java:91]: Overall, yes, retry util it's 1) interrupted by SIGTERM; 2) failure recovered by internal restart; 3) blocked if RocksDB thread goes to state D;

If RocksDB encounters an unrecoverable failure, for example, due to hardware failure, this thread will be blocked as the write opt prefers write-stall fast failure;
if RocksDB experiences recoverable failures,  sampling log error looks reasonable; 

Or, if you guys are suggesting something alternative, go ahead and comments are welcome

File: store/src/main/java/org/apache/rocketmq/store/queue/RocksDBConsumeQueueStore.java (hunk 3 above)
Comment (lizhanhui) [reply store/src/main/java/org/apache/rocketmq/store/queue/RocksDBConsumeQueueStore.java:540]: This override SHOULD not be included in the pull request, as it makes sense internally only.

File: broker/src/main/java/org/apache/rocketmq/broker/config/v2/ConfigStorage.java [hunk 4]
```diff
@@ -105,6 +141,15 @@ public byte[] get(ByteBuffer key) throws RocksDBException {
 
     public void write(WriteBatch writeBatch) throws RocksDBException {
         db.write(ableWalWriteOptions, writeBatch);
+        accountWriteOpsForWalFlush();
+    }
+
+    private void accountWriteOpsForWalFlush() throws RocksDBException {
+        int writeCount = writeOpsCounter.incrementAndGet();
+        if (writeCount >= messageStoreConfig.getRocksdbFlushWalFrequency()) {
+            this.db.flushWal(false);
```
[the remark this answers is not in the document]
Comment (lizhanhui) [reply broker/src/main/java/org/apache/rocketmq/broker/config/v2/ConfigStorage.java:150]: The frequency of commit-offset may be very high, I prefer to keep this periodic flush async.

File: store/src/main/java/org/apache/rocketmq/store/queue/RocksGroupCommitService.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+    public void putRequest(final DispatchRequest request) throws InterruptedException {
+        while (!buffer.offer(request, 3, TimeUnit.SECONDS)) {
+            log.warn("RocksGroupCommitService#buffer is full, 3s elapsed before space becomes available");
+        }
+        this.wakeup();
+    }
+
+    private void doCommit() {
+        boolean interrupted = false;
```
[the remark this answers is not in the document]
Comment (lizhanhui) [reply store/src/main/java/org/apache/rocketmq/store/queue/RocksGroupCommitService.java:72]: Good catch, will fix it in the next commit.

File: common/src/main/java/org/apache/rocketmq/common/config/AbstractRocksDBStorage.java [hunk 6]
```diff
@@ -121,14 +121,16 @@ protected void initWriteOptions() {
...
+        this.writeOptions.setNoSlowdown(false);
     }
 
     protected void initAbleWalWriteOptions() {
         this.ableWalWriteOptions = new WriteOptions();
         this.ableWalWriteOptions.setSync(false);
         this.ableWalWriteOptions.setDisableWAL(false);
-        this.ableWalWriteOptions.setNoSlowdown(true);
+        // https://github.com/facebook/rocksdb/wiki/Write-Stalls
+        this.ableWalWriteOptions.setNoSlowdown(false);
```
[the remark this answers is not in the document]
Comment (lizhanhui) [reply common/src/main/java/org/apache/rocketmq/common/config/AbstractRocksDBStorage.java:133]: Checkout the RocksGroupCommitService,  the group-commit thread is responsible for triggering and await batch-write to RocksDB; If the group-commit buffer[bounded, 100k, by default] is full, back-pressure the main dispatch thread, hence the dispatch-lag metrics alarms.

**Review by RongtongJin — APPROVED** [review]

**Review by fuyou001 — APPROVED** [review]

**Review by tianliuliu — DISMISSED** [review]

**Comment by lizhanhui** [comment]
> Tests in error: RocksdbTransferOffsetAndCqTest.testRocksdbCqWrite:147 NullPointer

Got a few pretty busy days, I would fix it when I got some bandwidth

**Review by RongtongJin — APPROVED** [review]

**Pull request merged**